pubsublite/operations.go

```diff
@@ -0,0 +1,152 @@
+
+func (p PublishTime) setRequest(req *pb.SeekSubscriptionRequest) {
+	req.Target = &pb.SeekSubscriptionRequest_TimeTarget{
+		TimeTarget: &pb.TimeTarget{
+			Time: &pb.TimeTarget_PublishTime{tspb.New(time.Time(p))},
+		},
+	}
+}
+
+// EventTime is a message event timestamp. It implements the SeekTarget
+// interface.
+type EventTime time.Time
+
+func (e EventTime) setRequest(req *pb.SeekSubscriptionRequest) {
+	req.Target = &pb.SeekSubscriptionRequest_TimeTarget{
+		TimeTarget: &pb.TimeTarget{
+			Time: &pb.TimeTarget_EventTime{tspb.New(time.Time(e))},
+		},
+	}
+}
+
+// SeekSubscriptionOption is reserved for future options.
+type SeekSubscriptionOption interface{}
+
+// OperationMetadata stores metadata for long-running operations.
+type OperationMetadata struct {
+	// The target of the operation. For example, targets of seeks are
+	// subscriptions, structured like:
+	// "projects/PROJECT_ID/locations/ZONE/subscriptions/SUBSCRIPTION_ID"
+	Target string
+
+	// The verb describing the kind of operation.
+	Verb string
+
+	// The time the operation was created.
+	CreateTime time.Time
+
+	// The time the operation finished running. Is zero if the operation has not
+	// completed.
+	EndTime time.Time
+}
+
+func protoToOperationMetadata(o *pb.OperationMetadata) (*OperationMetadata, error) {
+	if err := o.GetCreateTime().CheckValid(); err != nil {
+		return nil, err
+	}
+	metadata := &OperationMetadata{
+		Target:     o.Target,
+		Verb:       o.Verb,
+		CreateTime: o.GetCreateTime().AsTime(),
+	}
+	if o.GetEndTime() != nil {
+		if err := o.GetEndTime().CheckValid(); err != nil {
+			return nil, err
+		}
+		metadata.EndTime = o.GetEndTime().AsTime()
+	}
+	return metadata, nil
+}
+
+// SeekSubscriptionOperation manages a long-running seek operation from
+// AdminClient.SeekSubscription.
+type SeekSubscriptionOperation struct {
+	op *vkit.SeekSubscriptionOperation
+}
+
+// Name returns the path of the seek operation, in the format:
+// "projects/PROJECT_ID/locations/LOCATION/operations/OPERATION_ID".
+func (s *SeekSubscriptionOperation) Name() string {
+	return s.op.Name()
+}
+
+// Done returns whether the seek operation has completed.
+func (s *SeekSubscriptionOperation) Done() bool {
+	return s.op.Done()
+}
+
+// Metadata returns metadata associated with the seek operation. To get the
+// latest metadata, call this method after a successful call to Wait.
+func (s *SeekSubscriptionOperation) Metadata() (*OperationMetadata, error) {
+	m, err := s.op.Metadata()
+	if err != nil {
+		return nil, err
+	}
+	return protoToOperationMetadata(m)
+}
+
+// Wait polls until the seek operation is complete and returns one of the
+// following:
+//  - nil if the operation is complete and succeeded.
+//  - failure reason if the operation is complete and failed.
+//  - error if polling the operation status failed due to a non-retryable error.
+func (s *SeekSubscriptionOperation) Wait(ctx context.Context) error {
+	_, err := s.op.Wait(ctx)
+	return err
+}
```

Comment:
It looks like you have matched the LRO pattern from the gapic layer, which I agree seem right to me. Any reason to not also expose Poll? Then this would be more like every other LRO.

Reply:
I wasn't sure how often `Poll` is used in practice. The integration tests just call `Wait` and I noticed that the other handwritten libraries tend to just call `Wait` internally as well. So `Poll` was omitted to minimize the API surface, until a user asks for it.

Thanks for reviewing!

Reply:
SGTM. I think the main use case is when programming in a "single threaded" environment where you want a call that is not long-blocking. It would allow users to get a quick status check as well as metadata for the operation even if the operation is not in a terminal state.